src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/ClusterAwareConnectionProxy.java

```diff
@@ -778,701 +231,68 @@ public JdbcConnection getCurrentConnection() {
-    if ("cluster-".equalsIgnoreCase(clusterKeyword)
-        || "cluster-ro-".equalsIgnoreCase(clusterKeyword)) {
-      return matcher.group(1) + ".cluster-"
-          + matcher.group(3); // always RDS cluster endpoint
-    }
-    return null;
-  }
-
-  private boolean isRdsClusterDns(String host) {
-    Matcher matcher = auroraDnsPattern.matcher(host);
-    String clusterKeyword = getClusterKeyword(matcher);
-    return "cluster-".equalsIgnoreCase(clusterKeyword)
-        || "cluster-ro-".equalsIgnoreCase(clusterKeyword);
-  }
-
-  private boolean isReaderClusterDns(String host) {
-    Matcher matcher = auroraDnsPattern.matcher(host);
-    return "cluster-ro-".equalsIgnoreCase(getClusterKeyword(matcher));
-  }
-
-  private boolean isRdsCustomClusterDns(String host) {
-    Matcher matcher = auroraCustomClusterPattern.matcher(host);
-    return matcher.find();
-  }
-
-  private String getClusterKeyword(Matcher matcher) {
-    if (matcher.find()
-        && matcher.group(2) != null
-        && matcher.group(1) != null
-        && !matcher.group(1).isEmpty()) {
-      return matcher.group(2);
-    }
-    return null;
-  }
-
-  @Override
-  public synchronized Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
-    this.invokeStartTimeMs =
-        this.gatherPerfMetricsSetting ? System.currentTimeMillis() : 0;
-
-    Object result = super.invoke(proxy, method, args);
-
-    if (METHOD_CLOSE.equals(method.getName())) {
-      if (this.gatherPerfMetricsSetting) {
-        this.metrics.reportMetrics(this.log);
-        if (this.topologyService instanceof CanCollectPerformanceMetrics) {
-          ((CanCollectPerformanceMetrics) this.topologyService).reportMetrics(this.log);
-        }
-      }
-    }
-
-    return result;
-  }
-
-  /**
-   * Local method invocation handling for this proxy.
-   * This is the continuation of MultiHostConnectionProxy#invoke(Object, Method, Object[]).
-   */
-  @Override
-  protected synchronized Object invokeMore(Object proxy, Method method, Object[] args)
+  public synchronized Object invoke(Object proxy, Method method, Object[] args)
       throws Throwable {
-    final String methodName = method.getName();
+    String methodName = method.getName();
 
-    updateTopologyAndConnectIfNeeded(false);
+    if (METHOD_EQUALS.equals(methodName)) {
+      // Let args[0] "unwrap" to its InvocationHandler if it is a proxy.
+      return args[0].equals(this);
+    }
 
-    if (this.isClosed && !allowedOnClosedConnection(method)) {
-      invalidInvocationOnClosedConnection();
+    if (METHOD_HASH_CODE.equals(methodName)) {
+      return this.hashCode();
     }
 
-    Object result = null;
     try {
-      result = this.pluginManager.execute(
-          this.thisAsConnection.getClass(),
-          methodName,
-          () -> method.invoke(this.thisAsConnection, args));
-      result = proxyIfReturnTypeIsJdbcInterface(method.getReturnType(), result);
+      return invokeMore(proxy, method, args);
```

Comment:
We still need to wrap up results of invokeMore()

result = proxyIfReturnTypeIsJdbcInterface(method.getReturnType(), result);